### ISSUE TITLE
Function evaluation without configuration splitting

### ISSUE THREAD

**Issue by yzhang90** [kframework/kore#1065]
Here is how we evaluate functions:

```
@virgil-serbanuta : if we evaluate `sigma(f(x), g(y))`, we first evaluate `f(x)` to `F(x) /\ Pf(x)` and `g(y)` to `G(y) /\ Pg(y)`.
We will then evaluate `sigma(F(x), G(y))`. The child predicate while evaluating this is `Pf(x)` and `Pg(y)`.
so then, if `Pt(x, y)` is the top-level predicate, and we evaluate `sigma(F(x), G(y))` to `S(x, y) /\ Ps(x, y)`.
We will send `Ps(x, y)` to z3 with `Pt(x, y)` and `Pf(x)` and `Pg(y)` as a side predicate. 
We will and the main predicate and the side one before sending them to z3.
```
But the current approach doesn't fully take the advantage of passing top level predicate down and it will still split the configuration when doing function evaluation.

@ttuegel and I have the following proposal (also related to #945 ):
suppose we have two function eval rules:
```
rule f(x) => f1(x) requires x >=0
rule f(x) => f2(x) requires x<=0
```
the current config is `C(f(x)) /\ \top`.
`f(x)` can potentially have two results `f1(x) /\ x >=0` or `f2(x) /\ x <0`, when passing the top level predicate down, we check the validity of  `top -> x >=0` and `top -> x < 0`. In this example, none of the validity holds and then we still return `C(f(x)) /\ top` as the result and keep internally two possible outcomes of f(x) (using #945 ).
When semantics rules causes new constraint to be added to the configuration(config split), for example, the current configuration becomes `C(f(x)) /\ x >= 5`. Now, we know `x >= 5 -> x >=0`, and we can safely return `C(f1(x)) /\ x >= 5` as the result.

Suppose that the top level predicate is `Pt` and the side condition of the rule is `Ps`.
1. The function evaluation is different from applying semantics rules in the sense that:
For semantics rules, we should check if `Pt /\ Ps` is sat (if sat, we can apply the semantics rule);
For function rules, we should check if `Pt /\ not(Ps)` is unsat (if unsat, then Pt -> Ps and we can be sure that only this function rule can be applied and it will not cause the split of configuration)
2. Function evaluation can be localized. when evaluating `sigma(f(x), g(y))`, we first evaluate the arguments `f(x)` and `g(y)`. And when we evaluate `sigma`, we should only consider top level predicate `Pt` and we don't need to consider the child predicate generated from evaluating `f` and `g`.

**Comment by ttuegel** [comment]
> For function rules, we are checking if `Pt /\ not(Ps)` is unsat (if unsat, then Pt -> Ps and we can be sure that only this function rule can be applied and it will not cause the split of configuration)

We *should* check that `Pt ∧ ¬Ps` is unsat, but *do* we? 

> And when we evaluate `f(x)`, we should only consider top level predicate `Pt` and we don't need to consider the side condition when evaluating `sigma`.

I don't understand what "we don't need to consider the side condition when evaluating `sigma`" means. What side condition?

I'm not sure what is being proposed here. Is it to evaluate function rules using a different method than semantic rules? That seems to be what (1) suggests.

**Comment by yzhang90** [comment]
1. I changed the words: `we are checking` => `we should check`
2. I update the text. Based on what virgil describes, when evaluating `sigma`, we need to consider top level predicate `Pt(x, y)`, side condition `Ps(x, y)` for applying rules related to `sigma` and child predicates related to its arguments `Pf(x)` and `Pg(y)`. Now, we only need to consider `Pt(x, y)` and `Ps(x, y)`
3. yes. we should use different strategy to evaluate function rules than semantics rules.

**Comment by ttuegel** [comment]
> we need to consider top level predicate `Pt(x, y)`, side condition `Ps(x, y)` for applying rules related to `sigma` and child predicates `Pf(x)` and `Pg(y)` related to its arguments.  In the new approach, we only need to consider `Pt(x, y)` and `Ps(x, y)`

To clarify, when you say we do not need to consider `Pf` and `Pg`, that is because they are both `\top`.

**Comment by yzhang90** [comment]
>> we need to consider top level predicate Pt(x, y), side condition Ps(x, y) for applying rules related to sigma and child predicates Pf(x) and Pg(y) related to its arguments. In the new approach, we only need to consider Pt(x, y) and Ps(x, y)

> To clarify, when you say we do not need to consider Pf and Pg, that is because they are both \top.

Yes. because when evaluating f(x), we already showed that Pt(x,y) -> Pf(x)

**Comment by ttuegel** [comment]
Supersedes: #793.

One useful comment there:

>This will break the evaluation of some definedness conditions; we will need to figure out a way around that.

**Comment by ttuegel** [comment]
Needs: #1066

**Comment by yzhang90** [comment]
> This will break the evaluation of some definedness conditions; we will need to figure out a way around that.

Suppose we have 
```
rule f(x) => x + 1 if x >=0
rule f(x) => x -1 if x < 0
```
The top level predicate is `true`, then in the proposed approach, we can not further evaluate `f(x)` and just keep `f(x)` in the term. Later, when we need to check the defineness of `f(x)`,  `f(x)` is more or less an uninterpreted function to the backend and the backend may get stuck. I think it this situation, we can require users to add defineness lemmas for `f(x)`.

**Comment by ttuegel** [comment]
> I think it this situation, we can require users to add defineness lemmas for `f(x)`.

That seems like a reasonable approach.

**Comment by virgil-serbanuta** [comment]
> > For function rules, we are checking if `Pt /\ not(Ps)` is unsat (if unsat, then Pt -> Ps and we can be sure that only this function rule can be applied and it will not cause the split of configuration)
> 
> We _should_ check that `Pt ∧ ¬Ps` is unsat, but _do_ we?
> 

I think we do...

**Comment by ttuegel** [comment]
1. [ ] Make the change to evaluate functions without splitting, see what breaks, and report back. (estimate: XS)
2. ???

**Comment by ttuegel** [comment]
As a stop-gap, after disabling configuration splitting, we can introduce an axiom attribute to allow certain function rules to split the configuration.

**Comment by virgil-serbanuta** [comment]
First issue: let's say that we have the following function:
```
positive(x) = true if x > 0
positive(x) = false if x <= 0
```

And let us assume that we try to evaluate this function in a predicate, `positive(y) == true and y > 0`.

Note that, although this obviously evaluates, in a single branch, to just `y > 0`, right now we can't tell that because we don't send the top predicate down, as a side condition, when evaluating the top predicate.

Even if we do send it, stuff that is still unevaluated in the top predicate may allow us to prune branches.

(imperfect) Possible solutions:
1. Figure out a safe way of sending the predicate (i.e. perhaps we shouldn't assume the thing we're trying to prove, perhaps we should)
2. Send the previous configuration predicate

**Comment by virgil-serbanuta** [comment]
Example:
https://github.com/kframework/kore/blob/master/src/main/k/working/functions/prove/positive-requires-spec.k

**Comment by virgil-serbanuta** [comment]
Also, this one:
https://github.com/kframework/kore/blob/master/src/main/k/working/functions/prove/positive-spec.k

**Comment by ttuegel** [comment]
@virgil-serbanuta I wrote [here](https://github.com/kframework/kore/issues/1318#issuecomment-561681000) how to deal properly with the interaction between top-level predicates and `requires`.

**Comment by virgil-serbanuta** [comment]
evaluation in predicates seems to actually be the only issue in our integration tests, with those two tests mentioned above being the only ones that fail, and they are designed specifically for testing evaluation in predicates.

**Comment by virgil-serbanuta** [comment]
@ttuegel I know that.

I still (sort of) have two questions:
1. Should I try to solve both issues at the same time, or is it feasible to solve them separately?
2. Your post is about `requires` in the context of function evaluation. One can also get functions in the predicate through rewriting rules, so we still need to solve this issue.

I don't care much about the answer to 1, but I would prefer solving the issues in two pieces, if it's reasonable.

About 2 - I guess we can do something similar to what you mentioned, but for rewriting rules. I.e. if `R` is the requires clause, and `P` is the previous configuration predicate, we simplify `R` using `P` as a side condition (which, in principle, may mean that we remove the redundant parts of `R`), then we merge `P` and `simplified(R)` and we simplify again.

**Comment by ttuegel** [comment]
> Should I try to solve both issues at the same time, or is it feasible to solve them separately?

Sorry, I think I misunderstood the problem at first. Yes, I think we should try to solve them separately. I think it should be safe to send predicates to the solver which are not fully evaluated, as long as we have applied any substitutions. Some unevaluated predicates may not be useful, but I think we will iterate the simplifier until the condition stabilizes, so it should work out (?)

Another option would be: during predicate simplification, keep separately the part of the condition that is fully-simplified and the part that is not, using the former as the top-level conditions for evaluating functions in the latter. Although this seems cleaner, I'm not sure it would make a difference in practice, and it would be more complicated.

**Comment by virgil-serbanuta** [comment]
I just discovered a new issue, presented below together with a possible fix. Does that seem reasonable?

Currently we assume that we fully evaluate terms, and that if we don't evaluate something, we will not be able to evaluate it in the future, unless that something changes (e.g. we apply a substitution). This is signaled with the "simplified" attribute. We use the simplified attribute to stop evaluation for terms containing it.

However, if we implement the proposal above, we may be able to evaluate stuff without any local change (e.g. if the top-level predicate changes). This means that we have to re-evaluate some terms regardless of how may times we already attempted to evaluate them.

The first thing we need to change is to stop marking as simplified terms that we were not able to evaluate (we could have a special case for things that we were not able to evaluate because of branching and for things that contain them, but it could become complex). This would break our simplification property test (things that we simplify are marked as simplified), and we would attempt to  resimplify some things each time we get a chance.

Note that this change might be non-trivial, since when simplifying any kind of pattern we need to check whether its subpatterns are simplified before marking it as simplified.

The second change we likely need is in `TermLike.simplifyInternal`, since it's likely that playing with the `simplified` bit may break some of its assumptions.
